### PR TITLE
OA-24: Configurable support for token-based authentication.

### PR DESCRIPTION
**Pull request by wluyima**
https://issues.openmrs.org/browse/OA-24

### CODE REVIEW

File: omod/src/main/java/org/openmrs/module/oauth2login/web/filter/OAuth2ServiceAccountFilter.java
Comment (wluyima) [inline omod/src/main/java/org/openmrs/module/oauth2login/web/filter/OAuth2ServiceAccountFilter.java:71]: @mks-d what do you think of this?

**Review by ibacher — APPROVED** [review]
This seems correct to me. I've just got a few questions and some very nitpicky comments that can largely be ignored.

File: README.md
Comment (deniger) [reply README.md:130]: It's a specific need for WSO2 APIM:
https://apim.docs.wso2.com/en/3.0.0/learn/api-gateway/passing-end-user-attributes-to-the-backend/passing-enduser-attributes-to-the-backend-using-jwt/

We can't use Bearer due to some conflict in the header

**Comment by deniger** [comment]
Hello @wluyima 
I tried to test this PR but I got this error at module startup:
```
2021-08-25 12:13:17,499 +0000 logLevel=INFO thread="Thread-10" category=org.openmrs.api Exiting method saveGlobalProperty
Exception in thread "Thread-10" java.lang.LinkageError: loader constraint violation: when resolving method "org.slf4j.impl.StaticLoggerBinder.getLoggerFactory()Lorg/slf4j/ILoggerFactory;" the class loader (instance of org/openmrs/module/ModuleClassLoader) of the current class, org/slf4j/LoggerFactory, and the class loader (instance of org/apache/catalina/loader/WebappClassLoader) for the method's defining class, org/slf4j/impl/StaticLoggerBinder, have different Class objects for the type org/slf4j/ILoggerFactory used in the signature
        at org.slf4j.LoggerFactory.getILoggerFactory(LoggerFactory.java:418)
        at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:357)
        at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:383)
        at org.openmrs.module.oauth2login.web.filter.OAuth2ServiceAccountFilter.<clinit>(OAuth2ServiceAccountFilter.java:40)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at java.lang.Class.newInstance(Class.java:442)
        at org.openmrs.module.web.WebModuleUtil.loadFilters(WebModuleUtil.java:547)
        at org.openmrs.web.Listener.performWebStartOfModules(Listener.java:682)
        at org.openmrs.web.Listener.performWebStartOfModules(Listener.java:612)
        at org.openmrs.web.Listener.startOpenmrs(Listener.java:251)
        at org.openmrs.web.WebDaemon$1.run(WebDaemon.java:42)
```

File: omod/src/main/java/org/openmrs/module/oauth2login/web/JwtUtils.java
Comment (wluyima) [reply omod/src/main/java/org/openmrs/module/oauth2login/web/JwtUtils.java:103]: I think by 'unwritten convention' we don't code against environments, this is why we have configuration options like global properties for admins to decide things

Comment (ibacher) [reply omod/src/main/java/org/openmrs/module/oauth2login/web/JwtUtils.java:103]: Oh, yeah, I wasn't meaning to suggest that we have some sort of `isProduction()` check or something. My real suggestion would be to enforce that the URL is an HTTPS URL unless some globably property was set to disable that check.

File: omod/src/main/java/org/openmrs/module/oauth2login/web/HttpUtils.java
Comment (wluyima) [reply omod/src/main/java/org/openmrs/module/oauth2login/web/HttpUtils.java:46]: getContent does extra stuff like applying any registered content handler but I guess in this case it has no extra effect since we don't need that, so to avoid the unnecessary typecast I guess it's better to use getInputStream

File: omod/src/main/java/org/openmrs/module/oauth2login/web/HttpUtils.java
Comment (wluyima) [reply omod/src/main/java/org/openmrs/module/oauth2login/web/HttpUtils.java:31]: Not all IDPs currently support that content type

**Comment by icrc-fdeniger** [comment]
@mks-d @ibacher  is it possible to accept this PR as-is ?
We will have to update it and we found another bug ( the update task is updating roles even if the related configuration is empty).
So to avoid conflicts it will be better to accept this PR and we will propose fixes ASAP
Thanks

**Comment by dkayiwa** [comment]
@icrc-fdeniger i have just given you write access to this repository, just in case you are blocked by simply looking for someone with merge privileges. 😊

**Comment by icrc-fdeniger** [comment]
Thanks @dkayiwa  for this. Will accept this PR as soon I have time to propose an improvement

File: omod/src/main/resources/config.xml
Comment (mks-d) [inline omod/src/main/resources/config.xml:36]: @wluyima can we settle on `camelCase` or `kebab-case` for the filter names?
```xml
<filter-name>oauth2-service-account-filter</filter-name>
```
?

File: omod/src/test/resources/Keycloak/oauth2.properties
Comment (mks-d) [inline omod/src/test/resources/Keycloak/oauth2.properties:8]: I actually like this file to look like the real deal, can we put a value that looks like a public key?

File: omod/src/test/java/org/openmrs/module/oauth2login/web/JwtUtilsTest.java
Comment (mks-d) [inline omod/src/test/java/org/openmrs/module/oauth2login/web/JwtUtilsTest.java:112]: Like you did for `OAuth2ServiceAccountFilterTest`, could you make it clear visually where is the setup section, where is the replay section and where is the verify section? Especially for bigger tests.

I literally comment these parts usually eg: https://github.com/openmrs/openmrs-module-attachments/blob/bd8f77ba063203a6898b33ef4a856c64444b9738/omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java#L156-L169

File: README.md
Comment (mks-d) [inline README.md:146]: Let's show a snippet of **oauth2.properties** that looks like the real deal, like it is done a little higher in the README.

File: omod/src/test/java/org/openmrs/module/oauth2login/web/JwtUtilsTest.java
Comment (wluyima) [reply omod/src/test/java/org/openmrs/module/oauth2login/web/JwtUtilsTest.java:112]: Some code is self explaining, adding extra comments might be unnecessary noise especially for tests e.g adding comments to a no arg constructor but if that's how you prefer it I can do that

File: omod/src/main/resources/config.xml
Comment (wluyima) [reply omod/src/main/resources/config.xml:36]: ok

**Review by mks-d — APPROVED** [review]

File: omod/src/test/resources/Keycloak/oauth2.properties
Comment (wluyima) [inline omod/src/test/resources/Keycloak/oauth2.properties:8]: See hee

File: README.md
Comment (wluyima) [reply README.md:146]: There is already one, I updated though here https://github.com/openmrs/openmrs-module-oauth2login/pull/9/files#r714058607

**Pull request merged**
